src/core/tsi/alts/handshaker/alts_handshaker_client.cc

```diff
@@ -358,6 +335,164 @@ static tsi_result make_grpc_call(alts_handshaker_client* c, bool is_start) {
+      if (head_ == nullptr) {
+        GPR_ASSERT(tail_ == nullptr);
+        head_ = node;
+        tail_ = node;
+      } else {
+        GPR_ASSERT(tail_ != nullptr);
+        node->next = head_;
+        GPR_ASSERT(head_->prev == nullptr);
+        head_->prev = node;
+        head_ = node;
+      }
+    }
+    while (outstanding_handshakes_ < max_outstanding_handshakes_ &&
+           tail_ != nullptr) {
+      gpr_log(GPR_DEBUG,
+              "start handshake "
+              "outstanding_handshakes_:%ld "
+              "head_:%p tail_:%p",
+              outstanding_handshakes_, head_, tail_);
+      GPR_ASSERT(outstanding_handshakes_ < max_outstanding_handshakes_);
+      continue_make_grpc_call(tail_->client, true /* is_start */);
+      Node* prev = tail_;
+      tail_ = tail_->prev;
+      grpc_core::Delete(prev);
+      outstanding_handshakes_++;
+    }
+    if (tail_ == nullptr) {
+      head_ = nullptr;
+    }
+  }
+
+  grpc_core::Combiner* combiner_ = grpc_combiner_create();
+  Node* head_ = nullptr;
+  Node* tail_ = nullptr;
+  size_t outstanding_handshakes_ = 0;
+  const size_t max_outstanding_handshakes_;
+};
+
+gpr_once g_queued_handshakes_init = GPR_ONCE_INIT;
+/* Using separate queues for client and server handshakes is a
+ * hack that's mainly intended to satisfy the alts_concurrent_connectivity_test,
+ * which runs many concurrent handshakes where both endpoints
+ * are in the same process; this situation is problematic with a
+ * single queue because we have a high chance of using up all outstanding
+ * slots in the queue, such that there aren't any
+ * mutual client/server handshakes outstanding at the same time and
+ * able to make progress. */
+HandshakeQueue* g_client_handshake_queue;
+HandshakeQueue* g_server_handshake_queue;
+
+void DoHandshakeQueuesInit(void) {
+  g_client_handshake_queue =
+      grpc_core::New<HandshakeQueue>(40 /* max outstanding handshakes */);
+  g_server_handshake_queue =
+      grpc_core::New<HandshakeQueue>(40 /* max outstanding handshakes */);
+}
+
+void EnqueueHandshakeAndMaybeStartSome(alts_grpc_handshaker_client* client,
+                                       bool is_client) {
+  gpr_once_init(&g_queued_handshakes_init, DoHandshakeQueuesInit);
+  if (is_client) {
+    g_client_handshake_queue->EnqueueHandshakeAndMaybeStartSome(client);
+  } else {
+    g_server_handshake_queue->EnqueueHandshakeAndMaybeStartSome(client);
+  }
+}
+
+};  // namespace
+
+/**
+ * Populate grpc operation data with the fields of ALTS handshaker client and
+ * make a grpc call.
+ */
+static tsi_result make_grpc_call(alts_handshaker_client* c, bool is_start) {
+  GPR_ASSERT(c != nullptr);
+  alts_grpc_handshaker_client* client =
+      reinterpret_cast<alts_grpc_handshaker_client*>(c);
+  if (is_start) {
+    EnqueueHandshakeAndMaybeStartSome(client, client->is_client);
+    return TSI_OK;
+  } else {
+    return continue_make_grpc_call(client, is_start);
```

Comment:
Why do not we treat first (`is_start = true`) and second (`is_start = false`) ALTS handshake message differently? i.e., why do not we insert the second request into the queue as well?

Reply:
The reason for this is just based on the goal for the meaning of the queue: the queue is meant not to necessarily limit the number of outstanding handshakes messages, but instead it's meant to limit the number of outstanding handshake RPCs (note that this has the same queuing behavior of the `MAX_CONCURRENT_STREAMS` http/2 setting).

Thus, after we want to gate the starting of new RPCs on available slots in the handshake queue. But once a handshake has been admitted to start by the queue, then we proceed with it as fast as we can.

Reply:
That makes sense to me (especially the part mimicking `MAX_CONCURRENT_STREAMS` in http/2 setting).